### ISSUE TITLE
k8s support

### ISSUE THREAD

**Issue by Firstyear** [kanidm/kanidm#282]
It would be good to support kanidm's docker image working in k8s, especially with good native configuration and deployment options.

**Comment by phozzy** [comment]
I saw from you Dockerfile that you use configuration file to provide settings to Kanidm. Although it is possible to use config-maps to mount configuration file in k8s, the recommended way is to use environment variables. Is Kanidm able to check for env as its configuration? If not, I tnink it is the best point to start moving to k8s.

**Comment by phozzy** [comment]
The second point is about the support of clusterisation. Since main points of using k8s are: scaling in/out, rolling new releases or rolling back. And again from Dockerfile I see that `/data` volume is used for persistant data storage. That makes me think that the easiest way to support clusterisation is to do it on "storage" level. As an example of a good solution is one of storage backends for Hashikorp Vault: https://www.vaultproject.io/docs/configuration/storage/raft
Well... it is if you don't have any objections against raft.

**Comment by phozzy** [comment]
I think each of these points is worthy a dedicated issue. Should I open them?

**Comment by Firstyear** [comment]
The environment config yes. Storage level repl is not possible with kani, but I'll have to write something longer to explain that. So don't open that one up :)

**Comment by Firstyear** [comment]
https://github.com/kanidm/kanidm/blob/master/FAQ.md

I hope this explains why we can't use raft :)

**Comment by phozzy** [comment]
so #290 is here.

**Comment by phozzy** [comment]
> https://github.com/kanidm/kanidm/blob/master/FAQ.md
> 
> I hope this explains why we can't use raft :)

I checked that. I may be wrong, but is there a way to provide a full CAP for clustered environments?
You can also enable some new github features like "discussions" in the team/project settings to discuss such a questions there to prevent flooding here in issues.

**Comment by Firstyear** [comment]
No, CAP theorem states you have to choose two of the three. Partition tolerance is a requirement for kani so that leaves a choice of consistent or available. Consistent doesnt scale performance wise to a global scale system, but available does. And with a well modeled statemachone for conflict management, you can look pretty close to something thats all three (but isn’t). 389ds is AP, and helped inspire some parts of repl that I want for kani.

Another detail is for me this project is about getting things right rather than getting them done quick, so id rather takr longer and develop replication with extensive testing, modeling and simulation of time/partitioning etc, than rush and rely on a third party :)

William

**Comment by phozzy** [comment]
Actually the easiest way to check readiness of an application for dynamic environments such as k8s or similar is to check how well(good?) do it align with [12 factor app principles](https://12factor.net/). No jokes, those principles are kind of "Best practices".
Of course there is no need to follow them as strict as possible, but it is a good idea to keep them in mind as reference points.

**Comment by pando85** [comment]
I'm totally accord with that. We have to focus in converge in those principles.

**Comment by Firstyear** [comment]
Most 12 factor apps outsource their state to an external database or storage system that does the work to keep them in sync. In kani that's *not* possible, we are the provider of state, we are the provider of resources, and we must manage that correctly. So while it may be good to aim for, it also may not be possible. :(

**Comment by pando85** [comment]
Yep, in the future probably we could have an operator to manage kanidm replicas and scalability. It's always more complex because we have to manage state, backups and cluster configuration but it is good to get closer.

**Comment by sbordeyne** [comment]
I'm digging up an old thread, but I'm starting this repo for a KanIDM operator on k8s (for my own deployments) https://github.com/sbordeyne/kanidm-operator 

Might be useful in the future once feature complete, the only thing is that I'll write it in python for the ease of developpment (I am not proficient enough in Rust to use it in that fashion yet). Once released, that may be added to the official documentation (maybe in a "third-party tooling" section with a disclaimer that it's not maintained by the kanidm developers ?)

**Comment by pando85** [comment]
@sbordeyne I have that in mind since some years ago but I've never started with that.

For me, Go is clearly superior for this kind of projects. I clearly prefer rust syntax and I know better Python than Go but I have not doubts that Python is a worse decision than Go or Rust.

If you change the language probably I will join you at some point.

**Comment by yaleman** [comment]
I'm interested to see how this goes - if you want to use pyKanidm and want any support added just reach out. I should release the current code soon, but happy to put in extra before that happens.

**Comment by sbordeyne** [comment]
@yaleman Thanks for the reach out. I didn't know pyKanidm existed, it'll definitely speed up development. Is it the package available as `kanidm` on pypi.org ? It would be simpler for my build pipeline to not have to clone / build from source :smile: 

For now I do plan on handling some admin management tasks (Group, User, ServiceAccount, OAuthClient) and I planned on using the kanidm rust CLI for that (as a dirty hack, I'm working on that for my homelab setup which is currently, well, not accessible since I don't have an auth mechanism setup anymore since I switched over)

**Comment by yaleman** [comment]
Yeah, it's [available on pypi](https://pypi.org/project/kanidm/) but the dev version's available with `pip install -e git+https://github.com/kanidm/kanidm#subdirectory=pykanidm`. I'll try and remember to do a release this week 😇

**Comment by sbordeyne** [comment]
@yaleman I've perused through the KanidmClient class in pykanidm, and I note that there are a lot of features seemingly missing:

- create users
- create groups
- create service accounts
- bind service account access tokens
- list oauth2 clients
- create oauth2 client
- update methods (for the objects)
- delete methods for these objects

For now the client seems to only handle the auth part, maybe a bit of doc on the actual API behind could help in finishing up the client ? Or I could use the client by calling _call_post manually (and _call_put, _call_delete as needed) ?

**Comment by yaleman** [comment]
Yeah it's fairly limited because I've mainly only used it for the RADIUS integration and some testing, but it'll be easy to extend since you'd be writing the code either way

**Comment by yaleman** [comment]
I'll look at adding them this weekend unless you get there first 😄

**Comment by wrenix** [comment]
Hmm, for "storage" it is clear - beside sqlite a postgresql support is needed  - so no PV is needed.


Afterwards, a easy configuration ala the file provisioner in grafana is needed:
https://grafana.com/docs/grafana/latest/administration/provisioning/#data-sources

So an easy https://github.com/kiwigrid/k8s-sidecar could but them into the container and an easy configuration for Application (oauth2), Groups is possible (like in authentik and his blueprints):
https://github.com/goauthentik/authentik/issues/4119

**Comment by Firstyear** [comment]
@wrenix We actually removed all references to sql in our docs because this is a misnomer. You *can not* and *will never* be able to swap sqlite with another backend.

we are using sqlite as a dumb key-value store. It could be a CSV for all it matters.

Kanidm implements a full database, caching, transactions and more on top. As a result each server has in memory caches.

If you have two nodes point at one pgsql instance, then you either corrupt everything (since the two nodes memory isn't synced) or you destroy performance (because we can't cache and we are limited to pgsql). 

You must always consider kanidm as needing a persistent volume, and must always consider it as needing it's own replication. There are no external databases to add.

**Comment by yaleman** [comment]
Even with postgresql you'd need a PV at some point... 🤔

**Comment by wrenix** [comment]
@Firstyear hmm, okay, so kanidm would never be scaleable? :(

@yaleman it depends on your server setup - a cockroach as example (for an HA postgresql) - has an statefulset with many pvc's ...

**Comment by yaleman** [comment]
Please understand the problem before designing an orthogonal solution 😁 If you can find scaling limitations with the existing implementation, based on actual testing rather than assumptions then we are happy to discuss them.

**Comment by sbordeyne** [comment]
For what matters SQLite is actually very performant at what it does. Kanidm can scale horizontally, as long as you only need one write replica (otherwise you could end up with inconsistent db states I think). Backing up the data is also batteries included.

**Comment by Firstyear** [comment]
> For what matters SQLite is actually very performant at what it does. Kanidm can scale horizontally, as long as you only need one write replica (otherwise you could end up with inconsistent db states I think). Backing up the data is also batteries included.

No? You can have multiple writeable replicas.

Kanidm natively includes distributed eventually consistent replication.

**Comment by Firstyear** [comment]
> @Firstyear hmm, okay, so kanidm would never be scaleable? :(

https://kanidm.github.io/kanidm/stable/repl/index.html

**Comment by wrenix** [comment]
Then current setup option of repl is to static, to work well in k8s.
But maybe the Proposel of an coordinator could solve it (#2492):
https://kanidm.github.io/kanidm/master/developers/designs/replication_coord.html

My idea to simplify that it headless-service (DNS) and service (TCP-Loadbalance - sync between meshes/locations ) instatt was not welcome.

**Comment by yaleman** [comment]
Load-balancing of frontend nodes is fine, we're just enforcing tighter controls around the replication side of things. I could see a future state where k8s could manage the configuration of the RC through Custom Resource Definitions, but that's outside the current design scope; mainly because we only have so many people *doing* the work - rather than just offering suggestions on what to do 😄.

**Comment by Firstyear** [comment]
> My idea to simplify that it headless-service (DNS) and service (TCP-Loadbalance - sync between meshes/locations ) instatt was not welcome.

It's not that it wasn't welcome, is that every design decision we make has trade offs and consequences. We need to be able to test these designs, there are possible security weaknesses, and there is always a question of usability and supportability from our side. Another consideration is replication efficiency - network bandwidth between nodes is not unlimited and simple "full mesh" won't scale very far, introducing possible replication latency. 

We also have to balance that everyone has different requirements, each of which pull us in a different direction. 

In addition, this isn't my first time being involved with design and implementation of distributed databases. I have first hand seen a lot of good/bad from my experience with LDAP prior to this, and how those impacted people. 

Something you may not have seen is that this feature and planning has been going on for more than a year, with many discussions, ideas and choices made and refined based on new input from people. We need to not only make sure that replication coordination maintains connectivity between nodes without over-meshing that introduces latency, but we also need to ensure that changes to the topology can be reflected on nodes, and that the processes to join new nodes is secure since ability to replicate the kani database is a highly privileged action.

**Comment by yaleman** [comment]
I'm of the understanding that statefulsets would be required given the replication requirements, too...

**Comment by henobi** [comment]
Hey @yaleman and @pando85,
thank you for the feedback. That is a really good hint.
@jviets and I updated it from Deployment to StatefulSets.

For any further feedback please let us know.

**Comment by sfelde** [comment]
Hello, some updates here regarding k8n support? Thank you!
